nvme.c

```diff
@@ -5373,6 +5454,676 @@ static int show_registers(int argc, char **argv, struct command *cmd, struct plu
 }
 
+int get_reg_size(int offset)
+{
+	return nvme_is_64bit_reg(offset) ? sizeof(uint64_t) : sizeof(uint32_t);
+}
+
+static inline bool is_set_reg(struct get_reg_config *cfg, int offset)
+{
+	switch (offset) {
+	case NVME_REG_CAP:
+		if (cfg->cap)
+			return true;
+		break;
+	case NVME_REG_VS:
+		if (cfg->vs)
+			return true;
+		break;
+	case NVME_REG_INTMS:
+		if (cfg->intms)
+			return true;
+		break;
+	case NVME_REG_INTMC:
+		if (cfg->intmc)
+			return true;
+		break;
+	case NVME_REG_CC:
+		if (cfg->cc)
+			return true;
+		break;
+	case NVME_REG_CSTS:
+		if (cfg->csts)
+			return true;
+		break;
+	case NVME_REG_NSSR:
+		if (cfg->nssr)
+			return true;
+		break;
+	case NVME_REG_AQA:
+		if (cfg->aqa)
+			return true;
+		break;
+	case NVME_REG_ASQ:
+		if (cfg->asq)
+			return true;
+		break;
+	case NVME_REG_ACQ:
+		if (cfg->acq)
+			return true;
+		break;
+	case NVME_REG_CMBLOC:
+		if (cfg->cmbloc)
+			return true;
+		break;
+	case NVME_REG_CMBSZ:
+		if (cfg->cmbsz)
+			return true;
+		break;
+	case NVME_REG_BPINFO:
+		if (cfg->bpinfo)
+			return true;
+		break;
+	case NVME_REG_BPRSEL:
+		if (cfg->bprsel)
+			return true;
+		break;
+	case NVME_REG_BPMBL:
+		if (cfg->bpmbl)
+			return true;
+		break;
+	case NVME_REG_CMBMSC:
+		if (cfg->cmbmsc)
+			return true;
+		break;
+	case NVME_REG_CMBSTS:
+		if (cfg->cmbsts)
+			return true;
+		break;
+	case NVME_REG_CMBEBS:
+		if (cfg->cmbebs)
+			return true;
+		break;
+	case NVME_REG_CMBSWTP:
+		if (cfg->cmbswtp)
+			return true;
+		break;
+	case NVME_REG_NSSD:
+		if (cfg->nssd)
+			return true;
+		break;
+	case NVME_REG_CRTO:
+		if (cfg->crto)
+			return true;
+		break;
+	case NVME_REG_PMRCAP:
+		if (cfg->pmrcap)
+			return true;
+		break;
+	case NVME_REG_PMRCTL:
+		if (cfg->pmrctl)
+			return true;
+		break;
+	case NVME_REG_PMRSTS:
+		if (cfg->pmrsts)
+			return true;
+		break;
+	case NVME_REG_PMREBS:
+		if (cfg->pmrebs)
+			return true;
+		break;
+	case NVME_REG_PMRSWTP:
+		if (cfg->pmrswtp)
+			return true;
+		break;
+	case NVME_REG_PMRMSCL:
+		if (cfg->pmrmscl)
+			return true;
+		break;
+	case NVME_REG_PMRMSCU:
+		if (cfg->pmrmscu)
+			return true;
+		break;
+	default:
+		break;
+	}
+
+	return false;
+}
+
+static int get_register_properties(int fd, void **pbar, struct get_reg_config *cfg)
+{
+	int err;
+	int offset = NVME_REG_CRTO;
+	int size = offset + get_reg_size(offset);
+	__u64 value;
+	void *bar = malloc(size);
+	struct nvme_get_property_args args = {
+		.args_size = sizeof(args),
+		.fd = fd,
+		.value = &value,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+	};
+
+	if (!bar) {
+		nvme_show_error("malloc: %s", strerror(errno));
+		return -1;
+	}
+
+	memset(bar, 0xff, size);
+
+	for (offset = NVME_REG_CAP; offset <= NVME_REG_CRTO; offset += get_reg_size(offset)) {
+		if (cfg->offset != offset && !is_set_reg(cfg, offset))
+			continue;
+		args.offset = offset;
+		err = nvme_get_property(&args);
+		if (nvme_status_equals(err, NVME_STATUS_TYPE_NVME, NVME_SC_INVALID_FIELD)) {
+			value = -1;
```

Comment:
How do we distinguish from registers that were successfully read but had a value that was all 1s?

Seems like it would be sufficient to just skip non-fabrics registers?

Reply:
Right so just added to skip non-fabrics registers but still remained the invalid field erro checking since the fabrics registers also possible to be the error by the property get command.